### PR TITLE
[Tune] Add `Tuner.can_restore(path)` utility for checking if an experiment exists at a path/uri

### PR DESCRIPTION
**Pull request by justinvyu**
## Why are these changes needed?

This PR adds a utility to check if a given path (either local or remote) exists and can be restored from. It includes some simple validation that this is the root of the experiment directory (can't restore from the trial level directory).

Example workflow that this should be used in:

```python
import os
from ray.tune import Tuner
from ray.air import RunConfig

def train_fn(config):
    # Make sure to implement checkpointing so that progress gets
    # saved on restore.
    pass

name = "exp_name"
local_dir = "~/ray_results"
exp_dir = os.path.join(local_dir, name)

if Tuner.can_restore(exp_dir):
    tuner = Tuner.restore(exp_dir, resume_errored=True)
else:
    tuner = Tuner(
        "PPO",
        run_config=RunConfig(name=name, local_dir=local_dir),
    )
tuner.fit()
```

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tune/impl/tuner_internal.py [hunk 1]
```diff
@@ -298,6 +302,26 @@ def _validate_overwrite_trainable(
...
+        Returns:
+            can_restore: Whether or not this path exists and contains the pickled Tuner,
+                to load on restore.
+        """
+        path = str(path)
+        if is_non_local_path_uri(path):
+            dir_contents = list_at_uri(path)
+        else:
+            dir_contents = [] if not os.path.exists(path) else os.listdir(path)
+        return bool(dir_contents) and _TUNER_PKL in dir_contents
```
Comment (Yard1) [inline python/ray/tune/impl/tuner_internal.py:323]:
```suggestion
        return dir_contents and _TUNER_PKL in dir_contents
```

File: python/ray/tune/tests/test_tuner_restore.py [hunk 2]
```diff
@@ -964,6 +964,36 @@ def get_checkpoints(experiment_dir):
...
+    Tuner(
+        lambda config: None,
+        run_config=RunConfig(
+            name=name,
+            local_dir=str(tmpdir),
+            sync_config=tune.SyncConfig(upload_dir=upload_dir),
+        ),
+        tune_config=TuneConfig(trial_dirname_creator=lambda t: "trial_dir"),
+    ).fit()
+    assert Tuner.can_restore(path)
```
Comment (Yard1) [inline python/ray/tune/tests/test_tuner_restore.py:987]: can we also `assert not Tuner.can_restore(tmpdir)`?

**Review by Yard1 — APPROVED** [review]
LGTM!

File: python/ray/tune/tests/test_tuner_restore.py [hunk 3]
```diff
@@ -964,6 +964,41 @@ def get_checkpoints(experiment_dir):
...
 
+@pytest.mark.parametrize("upload_dir", [None, "memory:///test/"])
+def test_tuner_can_restore(tmp_path, upload_dir):
+    """Make sure that `can_restore` detects an existing experiment at a
+    local/remote path and only returns True if it's at the experiment dir root.
+    """
+    name = "exp_name"
+    path = tmp_path / name
+    if upload_dir:
+        path = Path(upload_dir) / name
```
Comment (krfricke) [inline python/ray/tune/tests/test_tuner_restore.py:975]: Nit for readability

```suggestion

    if upload_dir:
        path = Path(upload_dir) / name
    else:
        path = tmp_path / name
```

**Review by krfricke — APPROVED** [review]
LGTM

**Pull request merged**